br/pkg/lightning/backend/local/localhelper.go

```diff
@@ -415,48 +427,58 @@ func (local *Backend) BatchSplitRegions(
 		return nil, nil, errors.Annotatef(err, "batch split regions failed")
 	}
 	var failedErr error
-	scatterRegions := newRegions
+	splitRegions := newRegions
+	// wait for regions to be split
 	backoffer := split.NewWaitRegionOnlineBackoffer().(*split.WaitRegionOnlineBackoffer)
 	_ = utils.WithRetry(ctx, func() error {
 		retryRegions := make([]*split.RegionInfo, 0)
-		for _, region := range scatterRegions {
+		for _, region := range splitRegions {
 			// Wait for a while until the regions successfully splits.
 			ok, err2 := local.hasRegion(ctx, region.Region.Id)
 			if !ok || err2 != nil {
-				failedErr = err2
-				if failedErr == nil {
-					failedErr = errors.Errorf("region %d not found", region.Region.Id)
+				if err2 == nil {
+					log.FromContext(ctx).Warn("split region failed", zap.Uint64("regionID", region.Region.Id))
+				} else {
+					log.FromContext(ctx).Warn("split region failed", zap.Uint64("regionID", region.Region.Id), zap.Error(err2))
 				}
 				retryRegions = append(retryRegions, region)
 				continue
 			}
-			if err = local.splitCli.ScatterRegion(ctx, region); err != nil {
-				failedErr = err
-				retryRegions = append(retryRegions, region)
-			}
 		}
 		if len(retryRegions) == 0 {
 			return nil
 		}
 		// if the number of becomes smaller, we can infer TiKV side really
 		// made some progress so don't increase the retry times.
-		if len(retryRegions) < len(scatterRegions) {
+		if len(retryRegions) < len(splitRegions) {
 			backoffer.Stat.ReduceRetry()
 		}
-		// the scatter operation likely fails because region replicate not finish yet
-		// pack them to one log to avoid printing a lot warn logs.
-		log.FromContext(ctx).Warn("scatter region failed", zap.Int("regionCount", len(newRegions)),
-			zap.Int("failedCount", len(retryRegions)), zap.Error(failedErr))
-		scatterRegions = retryRegions
+		log.FromContext(ctx).Warn("split region failed", zap.Int("regionCount", len(newRegions)),
+			zap.Int("failedCount", len(retryRegions)))
+		splitRegions = retryRegions
 		// although it's not PDBatchScanRegion, WaitRegionOnlineBackoffer will only
 		// check this error class so we simply reuse it. Will refine WaitRegionOnlineBackoffer
 		// later
-		failedErr = errors.Annotatef(berrors.ErrPDBatchScanRegion, "scatter region failed")
+		failedErr = errors.Annotatef(berrors.ErrPDBatchScanRegion, "split region failed")
 		return failedErr
 	}, backoffer)
 
+	if ctx.Err() != nil {
+		return region, newRegions, ctx.Err()
+	}
+
+	// scatter regions
+	failedErr = nil
+	scatterRegions := newRegions
+	for _, region := range scatterRegions {
+		err = local.ScatterRegion(ctx, region)
+		if err != nil && failedErr == nil {
+			failedErr = errors.Annotatef(berrors.ErrPDBatchScanRegion, "scatter region failed")
```

Comment:
this should be done inside scope of `split.NewWaitRegionOnlineBackoffer().(*split.WaitRegionOnlineBackoffer)`

Reply:
i intentionally made it out of scope. Why we want to have two retry loops for scatter regions.

Reply:
see this, the backoffer only retry on `ErrPDBatchScanRegion` error, too bad😅

https://github.com/pingcap/tidb/blob/abec0c68e368d6dd5c076f725c42813a426f8ccc/br/pkg/restore/split/split.go#L221-L235

Reply:
oh, should i change the Nextbackoffer. NextBackoff should retry on any error ideally. It should be the caller responsibility.

Reply:
as commented in here, we plan to do it later, but if you have time, you can refactor this part too
https://github.com/pingcap/tidb/blob/64a2e1c20a0666f8f8c4ae46330dedcc406a5d47/br/pkg/lightning/backend/local/localhelper.go#L421-L425